src/library/compiler/csimp.cpp

```diff
@@ -1136,7 +1136,6 @@ class csimp_fn {
     expr visit_lambda(expr e, bool is_join_point_def, bool top) {
         lean_assert(is_lambda(e));
         lean_assert(!top || m_fvars.size() == 0);
-        lean_assert(!top || !already_simplified(e));
```

Comment:
@leodemoura It's not clear to me why this one doesn't hold

Reply:
I will take a look. What is holding before?

Reply:
I don't know

Reply:
I investigated this issue. This assertion does not hold anymore. 
A few months ago, I added the following function that combines `csimp`, `cse` and dead `let` elimination in a loop until no further simplifications are possible. I reuse the simplifier object. Thus, an expression may be top-level and have already been simplified in the previous iteration.

Reply:
https://github.com/leanprover/lean4/blob/master/src/library/compiler/csimp.cpp#L2001